fix(form-builder): fix scroll to invalid input behaviour

orbitalsqwib opened the pull request:
## Problem

After submission of a form with missing or invalid fields, the checker will currently automatically focus on the offending field (and by extension, scroll it into view).

However, the preview nav bar is not taken into account when this occurs during the preview modal, and thus displays a different behaviour than expected, as the focused element is often hidden behind the nav bar.

This can lead to confusion on the user's part as to why the checker is not submitting as expected, and can be misconstrued as a bug.

Additionally, on published checkers, only the input of the offending field is scrolled into view, excluding the title and description. This requires users to scroll up slightly more to see what is the context behind the invalid field.

Closes #515

## Solution

**Features**:

- implemented virtual controller inputs to support `react-hook-form` scroll into view for irregular, non-standard inputs that consist of more than one input element, such as radio groups, checkbox fields and date selectors.

**Improvements**:

- Add a scroll container over the tab content, such that nav bars are not included as part of the scroll area.
- Add top scroll margins to virtual / non-virtual inputs to include field titles and descriptions on a best effort basis

## Before & After Screenshots

**BEFORE**:
Previewed checker
![Screenshot 2021-05-11 at 11 28 39 AM](https://user-images.githubusercontent.com/21305518/117756512-de33da00-b250-11eb-9251-47fd2a66a13f.png)

Published checker
![Screenshot 2021-05-11 at 11 28 26 AM](https://user-images.githubusercontent.com/21305518/117756516-e0963400-b250-11eb-88cc-39e81283898d.png)

**AFTER**:
Previewed checker
![Screenshot 2021-05-11 at 12 02 34 PM](https://user-images.githubusercontent.com/21305518/117756584-fe639900-b250-11eb-8988-74576a3b014a.png)

Published checker
![Screenshot 2021-05-11 at 12 02 57 PM](https://user-images.githubusercontent.com/21305518/117756594-03c0e380-b251-11eb-9a7f-faf5e163b986.png)

## Tests

- [ ] Check that published checker scrolls to invalid inputs as per above screenshots (title and description both visible)
- [ ] Check that previewed checker scrolls to invalid inputs as per above screenshots (should have exactly the same behaviour as published checkers)

Commit:
fix(form-builder): fix scroll container

- fix scroll insets by nesting tab container, such that scrollIntoView
  works as expected with nav bars.

Commit:
fix(fields): fix invalid input onfocus scroll offsets

- implement virtual controller inputs for irregular inputs such as
  radio groups, date and checkbox fields
- implement scroll margin offsets to capture field title and description
  on a best effort-basis

Commit:
refactor(virtual-controller-input): use external style prop

Commit:
Merge branch 'develop' into fix/invalid-input-scroll-behaviour

Commit:
fix(virtual-controller-input): prevent tab focus on virtual input fields

lamkeewei commented:
Was wondering if there's a simpler approach to solving this problem than to introduce a `VirtualControllerInput`. Not sure if there's any unintended side effects if the `ref` is not bounded to the actual input.

orbitalsqwib commented:
According to the react-hook-form documentation, the `field.ref` return value from `Controller` is only responsible for focusing the error inputs.

> A ref used to connect hook form to the input. Assign ref to component's input ref to allow hook form to focus the error input.

As such, I don't think there should be other side effects from assigning the ref to a virtual input, other than that it might disrupt navigation/accessibility.

I think I can give fixing the refs on problematic fields such as `DateField` and `RadioField` another shot, but if it doesn't pan out, the virtual input might be better alternative than having a broken scroll-to-focus

lamkeewei commented:
> According to the react-hook-form documentation, the `field.ref` return value from `Controller` is only responsible for focusing the error inputs.
> 
> > A ref used to connect hook form to the input. Assign ref to component's input ref to allow hook form to focus the error input.
> 
> As such, I don't think there should be other side effects from assigning the ref to a virtual input, other than that it might disrupt navigation/accessibility.
> 
> I think I can give fixing the refs on problematic fields such as `DateField` and `RadioField` another shot, but if it doesn't pan out, the virtual input might be better alternative than having a broken scroll-to-focus

Sure - I think worthwhile to take a second look. Just felt a little uneasy fixing a minor UI issue with a fairly non-standard change.

Commit:
fix(refs): fix refs on date, radio and checkbox fields

- remove virtual inputs
- override chakra class css for radio and checkbox fields

orbitalsqwib commented:
Managed to fix the refs across the troublesome fields. Turned out to be issues with how Chakra composes and styles the Radio and Checkbox inputs, as well as how Flatpickr manages the refs on the rendered input field.

Main notable changes are the two additional css files that add `scrollMarginTop` to the internal inputs on the Chakra Radio/Checkbox fields, which should be okay for our current use case, provided that we don't need specific scroll margins for each field.

lamkeewei commented on src/client/styles/checkbox-field-input.css:
```diff
@@ -0,0 +1,3 @@
+.chakra-checkbox__input {
+    scroll-margin-top: 88px;
```
Is it possible to override this value [using Chakra](https://chakra-ui.com/docs/theming/component-style#styling-single-part-components) instead of patching the CSS directly? How Chakra name their CSS classes should be transparent to us.

orbitalsqwib replied:
It might be possible to override the value through overriding the global default input themes, but I feel that this solution might be too invasive and cause unexpected behaviours on modals where we may want to customise the same set of styles on another page with different values.

As such, I'd still prefer that we just override the CSS directly, though we could discuss about this if you have strong opinions on the subject.

lamkeewei replied:
As discussed on Slack, this current way of overriding the CSS also affects _all_ checkbox inputs since it's overriding global styles.

Commit:
refactor(radio/checkbox-inputs): improve css targeting

- implement box wrapper to improve css targeting to only checker field
  checkbox or radio inputs
- clean up and document css files

lamkeewei reviewed (approved):
lgtm!
